apps/aecore/lib/aecore/peers/sync.ex

```diff
@@ -112,5 +147,60 @@ defmodule Aecore.Peers.Sync do
       end
     end)
   end
-end
 
+  # Builds a chain, starting from the given block,
+  # until we reach a block, of which the previous block is the highest in our chain
+  # (that means we can add this chain to ours)
+  defp build_chain(state, block, chain) do
+    has_parent_block_in_state = Map.has_key?(state, block.header.prev_hash)
+    has_parent_in_chain =
+      block.header.prev_hash == BlockValidation.block_header_hash(Chain.latest_block().header)
+    cond do
+      has_parent_block_in_state ->
+        build_chain(state, state[block.header.prev_hash], [block | chain])
+      has_parent_in_chain ->
+        [block | chain]
+      true ->
+        []
+    end
+  end
+
+  # Adds the given chain to the local chain and
+  # deletes the blocks we added from the state
+  defp add_built_chain(chain, state) do
+    Enum.reduce(chain, state, fn (block, acc) ->
+        case Chain.add_block(block) do
+          :ok ->
+            Map.delete(acc, BlockValidation.block_header_hash(block.header))
+          :error ->
+            acc
+        end
+      end)
+  end
+
+  # Gets all unknown blocks, starting from the given one
+  defp check_peer_block(peer_uri, block_hash, state) do
+    case Chain.has_block?(block_hash) do
+      false ->
+        case(HttpClient.get_block({peer_uri, block_hash})) do
+          {:ok, peer_block} ->
+            deserialized_block = Serialization.block(peer_block, :deserialize)
+            try do
+              BlockValidation.single_validate_block(deserialized_block)
+              peer_block_hash =
+                BlockValidation.block_header_hash(deserialized_block.header)
+
```

Comment:
We should check here if `peer_block_hash` matches `block_hash`.

Reply:
yes, great find!